### ISSUE TITLE
Request or help if exists already - `#js` reader syntax in cljs interop

### ISSUE THREAD

**Issue by riotrah** [kkinnear/zprint#292]
I write a lot of cljs, and thus a lot of js interop. This frequently involves use of `#js {:my-key ""}` and `#js [""]`

I'm not sure what it is about my config, which is otherwise really great, but whenever a hang or flow are warranted (line too long, or explicitly configured, etc), the `#js` ends up on its own line.

Is that expected behavior? Is there, or could there be, a config option to keep `#js` always adjacent (left of) the subsequent form, whilst maintaining the rest of the hang/flow?

**Comment by kkinnear** [comment]
I think I'm going to need your config and an actual example to make progress on this.  When I try it with the default config, it does exactly what you want:
```
(czprint i292 {:parse-string? true :width 50})
(defn i292
  [x y]
  (let [this-is (a-test this
                        is
                        only
                        a
                        test
                        #js {:my-key ""})]
    (vector? this-is)))
```
I checked, and it has worked like this for a good long time in the past.

Something in your configuration must be triggering the problem, so if you give me an example and your config, I'll figure out what it is and see what I can to do to prevent it.  Thanks!

**Comment by kkinnear** [comment]
Thanks for the configurations.  That didn't provoke the problem, interestingly enough.  It seems like, using the default configuration, that you get the `#js` on a line by itself when the map after it doesn't fit on the same line.  Like this:
```
(defn i292
  [x y]
  (let [this-is (a-test this
                        #js
                         {:your-key "",
                          :a-very-long-key :and-a-very-long-value,
                          :a-second-key :and-a-second-value}
                        a
                        test
                        #js {:my-key ""}
                        not
                        the
                        end)]
    (vector? this-is)))
```
Notice the two `#js` elements above.  

Questions:

  1. I think the first `#js` is what you don't like and the second is what you do like.  Is this correct?
  2. Is this the behavior that you see?   That is, are the places where `#js` appears on its own line places where the thing following it would not fit on the same line as the `#js`?
  3. Do you ever see things like the second `#js`, where the next thing stays on the same line as the `#js`?

Thanks!

**Comment by riotrah** [comment]
1. Yes, tho needs some upcoming context 
2. Yep
3. Yep! I neglected to even notice that as all of my usages of that recently have been for somewhat larger/nested pojos/maps

Your first example comes close, tho it's a more "justifiable" (?) case for a separate line `#js` , imo.

But here's a real case:

> _suppose there's a max width set somewhere near the end of the second line_

**Before:**

```clojure
				   #js {:a "b"}
                   #js {:detail (or signature (str "[" (name (or category "")) "]"))
                        :documentation (when description
                                         #js {:value description})
                        :insertText (if function? (str/upper-case insert-text) insert-text)
                        :insertTextRules (when snippet
                                           insert-as-snippet-rule)
                        :kind kind
                        :label s
                        :filterText (str/upper-case s)
                        :sortText (str/lower-case (str sort-text-prefix " " s))}))
```

**After zprint:**

```clojure
					 #js {:a "b"}
                     #js
                      {:detail (or signature (str "[" (name (or category "")) "]"))
                       :documentation (when description
                                        #js {:value description})
                       :insertText (if function? (str/upper-case insert-text) insert-text)
                       :insertTextRules (when snippet
                                          insert-as-snippet-rule)
                       :kind kind
                       :label s
                       :filterText (str/upper-case s)
                       :sortText (str/lower-case (str sort-text-prefix " " s))}))
```

I'm not sure how I'd configure zprint to do what my preferred output would look like:

**Preferred:**
```clojure
#js {:a 'b}
                   #js {:detail (or signature 
								    (str "[" (name (or category "")) "]"))
                        :documentation (when description
                                         #js {:value description})
                        :insertText (if function? 
									    (str/upper-case insert-text) 
                                        insert-text)
                        :insertTextRules (when snippet
                                           insert-as-snippet-rule)
                        :kind kind
                        :label s
                        :filterText (str/upper-case s)
                        :sortText (str/lower-case (str sort-text-prefix 
           										       " " s))}
```

The specifics of where exactly all of the newlines should be or when to hang/flow/etc are less relevant (for this demo at least) than the main point being to start "breaking" the lines _in_ the map _before_ considering placing the `#js` on its own line. Ie, I'd rather have a map with multiline rhs values before I'd have a map whose `#js` is a line above it.

Does that make sense?

**Comment by kkinnear** [comment]
Yes, it makes a lot of sense.  

It is worth knowing that you can't configure zprint to do it the way you want.  At least I haven't found any way to do that.  I am experimenting with a code change that will do it the way you want, which I think is probably what most people would want too.  I'm currently looking at how to get it to go back to the not-great way it does it now in case someone really wants that.

**Comment by riotrah** [comment]
Wow thanks so much for exploring this wow!!!!!!

**Comment by kkinnear** [comment]
In `1.2.6` the `#js` will try to put the next thing on the same line, as you want it, by default.  There will be a new configuration element in the option map: `:tagged-literal` which will be configurable in such a way that if someone wants it the way it is today, they can configure it that way by setting`{:tagged-literal {:hang? false}}`.  I'll let you know when `1.2.6` is available.

**Comment by kkinnear** [comment]
I just released `1.2.6`.  The changes I made to format tagged literals are the default, so you should see what you want without anything other than upgrading.  Thanks for pointing this out!
